File: apps/viewer-web/src/features/publications/isbn.ts
export function normalizeISBN13(value: string): string {
  const digits = Array.from(value.trim())
    .filter((character) => character >= "0" && character <= "9")
    .join("");
  if (digits.length !== 13) {
    return "";
  }
  if (!digits.startsWith("978") && !digits.startsWith("979")) {
    return "";
  }
  let sum = 0;
  for (let index = 0; index < 12; index += 1) {
    const digit = Number(digits[index]);
    sum += index % 2 === 0 ? digit : digit * 3;
  }
  const check = (10 - (sum % 10)) % 10;
  return check === Number(digits[12]) ? digits : "";
}

File: apps/viewer-api-go/internal/publications/isbn.go
package publications

import "strings"

func NormalizeISBN13(value string) string {
	digits := make([]rune, 0, 13)
	for _, r := range strings.TrimSpace(value) {
		if r >= '0' && r <= '9' {
			digits = append(digits, r)
		}
	}
	if len(digits) != 13 {
		return ""
	}
	isbn := string(digits)
	if !strings.HasPrefix(isbn, "978") && !strings.HasPrefix(isbn, "979") {
		return ""
	}
	if !validISBN13Checksum(isbn) {
		return ""
	}
	return isbn
}

func validISBN13Checksum(isbn string) bool {
	if len(isbn) != 13 {
		return false
	}
	sum := 0
	for i := 0; i < 12; i++ {
		digit := int(isbn[i] - '0')
		if i%2 == 0 {
			sum += digit
		} else {
			sum += digit * 3
		}
	}
	check := (10 - (sum % 10)) % 10
	return check == int(isbn[12]-'0')
}
